### ISSUE TITLE
Which native extensions should we package?

### ISSUE THREAD

**Issue by FooBarWidget** [phusion/traveling-ruby#6]
In issue #3 I have described how I envision support for native extensions to work. Since requiring developers to build native extensions themselves is too much hassle, I proposed providing precompiled versions of the most popular native extensions, that developers can just easily drop into their package.

But this raises the question: which native extensions? We can't automatically package them all. For reasons explained in #3, packaging native extensions requires a lot of manual work. So we have to choose.

I plan on packaging these gems:

Because they're popular:
- nokogori
- json

Because Rails apps likely need these:
- sqlite3
- mysql2
- pg
- bcrypt

I've opened a poll for more native extensions. Please add your input here: https://docs.google.com/spreadsheets/d/1oNgrwHyV9UaMGFTttsfA9QPvL461NWLfHeeAap3Az_A/edit?usp=sharing

**Comment by FooBarWidget** [comment]
RedCloth, escape_utils, posix-spawn, nokogumbo and github-markdown have been added.

**Comment by FooBarWidget** [comment]
rugged has been added.

**Comment by FooBarWidget** [comment]
charlock_holmes has been added.

**Comment by etdsoft** [comment]
Hey Hongli, do you plan to push a new release now that all the Octodown gems have been added?
We're planing to use traveling-ruby for the next release of [Dradis Framework](https://github.com/dradis/dradisframework) too. This is where all the RedCloth stuff was coming from.

**Comment by FooBarWidget** [comment]
Soon. I working on Traveling Ruby occasionally during the weekend, but during the week most of my time is occupied by Passenger maintenance/development as well as business operations.

**Comment by etdsoft** [comment]
Of course, makes sense, no rush.

**Comment by FooBarWidget** [comment]
The unf_ext gem has been added.

**Comment by FooBarWidget** [comment]
Traveling Ruby 20150130 is out. Next version is planned to use Ruby 2.2, as per GH-28.

**Comment by etdsoft** [comment]
excellent!

**Comment by sleewoo** [comment]
what about Puma?

**Comment by FooBarWidget** [comment]
Added puma, unicorn, kgio, raindrops, fast-stemmer, hitimes, redcarpet.

**Comment by sleewoo** [comment]
awesome!
thanks.

**Comment by sleewoo** [comment]
ok, i see `puma`, `unicorn` etc. in shared/gemfiles/20150210-next of `next` branch.
would they be available on s3 any soon?
thanks for your efforts.

**Comment by Nowaker** [comment]
I added [nio4r](https://github.com/celluloid/nio4r) to the list. This is a dependency of [Celluloid::IO](https://github.com/celluloid/celluloid-io) which is great for making concurrent HTTP requests.

**Comment by Nowaker** [comment]
@FooBarWidget The same as @sleewoo - when can you release binaries for puma? The spreadsheet says "Already packaged" but Puma is not available at http://traveling-ruby.s3-us-west-2.amazonaws.com/list.html.

**Comment by FooBarWidget** [comment]
It's packaged in the development branch. I'm sick right now and there are lots of high-priority non-Travling-Ruby things on my todo list, so I'm afraid I can't give you a time indication right now.

**Comment by Nowaker** [comment]
Are there any binaries for the development branch?

**Comment by FooBarWidget** [comment]
No. I build it from my laptop on every release. If you need binaries now, you should run the build scripts yourself.

**Comment by sleewoo** [comment]
i was able to build everything in next branch.
could provide output if interested.
extensions are full of backdoors of course :)

**Comment by michaelbarton** [comment]
Could it be possible to provide a native extension of gherkin, the gem that is used for cucumber? We would like to share our test suite for bioinformatics tools powered by cucumber. I gave this a try today but I saw that gherkin is a native extension.

**Comment by FooBarWidget** [comment]
I'll look into it. I've scheduled some time for Traveling Ruby maintenance work this week. Right now I'm blocked by https://bitbucket.org/ged/ruby-pg/issue/219/please-release-a-new-gem.

**Comment by michaelbarton** [comment]
Thanks. I took a look at the code for building native extensions. I think I almost got it to work, I was able to get everything to appear to compile but I wasn't sure about where the final output was produced, or if it was produced. I think with a little more documentation I think I might have been able to do this myself.

**Comment by FooBarWidget** [comment]
@michaelbarton The build output is in the 'output' directory. The 'package' script packages the output into tarballs.

Are you talking about gherkin or gherkin3?

**Comment by michaelbarton** [comment]
I think gherkin. From what I understand gherkin3 is still in development.

**Comment by FooBarWidget** [comment]
Ok, version 20150517 is released now, with support for many extensions requested in the past 2 months.

**Comment by dapicester** [comment]
I added [websocket-driver](https://github.com/faye/websocket-driver-ruby). It was extracted from [Faye](http://faye.jcoglan.com/) to provide WebSocket support for Ruby.

**Comment by ashes999** [comment]
Hi,

I added [gosu](http://libgosu.org/). Like most remaining gems on the list, it only has one vote (mine). If you can get back to me about whether this is feasible or not (not actually building the binaries, but assessing it), that would be great @FooBarWidget

**Comment by Nowaker** [comment]
I've added concurrent-ruby to the list. It's become a very important library now that Sprockets depend on it.
